File: c/monster_walk_primes.c
// C: Monster Walk with Prime Factorizations in All Bases

#include <stdio.h>
#include <stdint.h>
#include <gmp.h>

#define NUM_PRIMES 15

// Prime factorization structure
typedef struct {
    uint32_t primes[NUM_PRIMES];
    uint32_t exponents[NUM_PRIMES];
} PrimeFactorization;

// Monster primes
const uint32_t MONSTER_PRIMES[NUM_PRIMES] = {
    2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 41, 47, 59, 71
};

// Monster factorization
PrimeFactorization monster_factorization() {
    PrimeFactorization f;
    for (int i = 0; i < NUM_PRIMES; i++) {
        f.primes[i] = MONSTER_PRIMES[i];
    }
    f.exponents[0] = 46;  // 2^46
    f.exponents[1] = 20;  // 3^20
    f.exponents[2] = 9;   // 5^9
    f.exponents[3] = 6;   // 7^6
    f.exponents[4] = 2;   // 11^2
    f.exponents[5] = 3;   // 13^3
    for (int i = 6; i < NUM_PRIMES; i++) {
        f.exponents[i] = 1;
    }
    return f;
}

// Step 4: Remove 8 factors
PrimeFactorization step4_remaining() {
    PrimeFactorization f = monster_factorization();
    f.exponents[3] = 0;   // Remove 7^6
    f.exponents[4] = 0;   // Remove 11^2
    f.exponents[6] = 0;   // Remove 17
    f.exponents[7] = 0;   // Remove 19
    f.exponents[9] = 0;   // Remove 29
    f.exponents[10] = 0;  // Remove 31
    f.exponents[11] = 0;  // Remove 41
    f.exponents[13] = 0;  // Remove 59
    return f;
}

// Compute value from factorization
void compute_value(PrimeFactorization *f, mpz_t result) {
    mpz_set_ui(result, 1);
    mpz_t temp;
    mpz_init(temp);
    
    for (int i = 0; i < NUM_PRIMES; i++) {
        if (f->exponents[i] > 0) {
            mpz_ui_pow_ui(temp, f->primes[i], f->exponents[i]);
            mpz_mul(result, result, temp);
        }
    }
    
    mpz_clear(temp);
}

// Convert to base
void to_base(mpz_t n, uint32_t base, uint8_t *digits, size_t *len) {
    mpz_t temp;
    mpz_init_set(temp, n);
    
    *len = 0;
    while (mpz_cmp_ui(temp, 0) > 0) {
        digits[*len] = mpz_fdiv_ui(temp, base);
        mpz_fdiv_q_ui(temp, temp, base);
        (*len)++;
    }
    
    // Reverse
    for (size_t i = 0; i < *len / 2; i++) {
        uint8_t tmp = digits[i];
        digits[i] = digits[*len - 1 - i];
        digits[*len - 1 - i] = tmp;
    }
    
    mpz_clear(temp);
}

// Print factorization
void print_factorization(PrimeFactorization *f) {
    int first = 1;
    for (int i = 0; i < NUM_PRIMES; i++) {
        if (f->exponents[i] > 0) {
            if (!first) printf(" × ");
            if (f->exponents[i] == 1) {
                printf("%u", f->primes[i]);
            } else {
                printf("%u^%u", f->primes[i], f->exponents[i]);
            }
            first = 0;
        }
    }
}

int main() {
    printf("🔢 Monster Walk with Primes - All Bases (C)\n");
    printf("============================================\n\n");
    
    mpz_t monster_val, step4_val;
    mpz_init(monster_val);
    mpz_init(step4_val);
    
    // Step 1: Full Monster
    PrimeFactorization monster = monster_factorization();
    compute_value(&monster, monster_val);
    
    printf("Step 1: Full Monster\n");
    printf("  Primes: ");
    print_factorization(&monster);
    printf("\n");
    printf("  Decimal: ");
    mpz_out_str(stdout, 10, monster_val);
    printf("\n");
    printf("  Hex: 0x");
    mpz_out_str(stdout, 16, monster_val);
    printf("\n\n");
    
    // Step 4: Remove 8 factors
    PrimeFactorization step4 = step4_remaining();
    compute_value(&step4, step4_val);
    
    printf("Step 4: Remove 8 factors (Group 1) ⭐\n");
    printf("  Remaining: ");
    print_factorization(&step4);
    printf("\n");
    printf("  Decimal: ");
    mpz_out_str(stdout, 10, step4_val);
    printf("\n");
    printf("  Hex: 0x");
    mpz_out_str(stdout, 16, step4_val);
    printf("\n\n");
    
    // Test in all bases
    printf("Testing all bases 2-71:\n");
    for (uint32_t base = 2; base <= 71; base++) {
        uint8_t digits[256];
        size_t len;
        to_base(step4_val, base, digits, &len);
        
        if (base == 2 || base == 10 || base == 16 || base == 71) {
            printf("  Base %2u: %zu digits\n", base, len);
        }
    }
    
    printf("\n✅ All bases computed\n");
    
    mpz_clear(monster_val);
    mpz_clear(step4_val);
    
    return 0;
}

File: src/bin/monster_walk_primes.rs
// Rust: Monster Walk with Prime Factorizations in All Bases

use num_bigint::BigUint;
use num_traits::{One, Zero, ToPrimitive};
use std::collections::HashMap;

/// Prime factorization
#[derive(Debug, Clone)]
struct PrimeFactorization {
    factors: HashMap<u32, u32>,
}

impl PrimeFactorization {
    fn new() -> Self {
        Self { factors: HashMap::new() }
    }
    
    fn monster() -> Self {
        let mut f = Self::new();
        f.factors.insert(2, 46);
        f.factors.insert(3, 20);
        f.factors.insert(5, 9);
        f.factors.insert(7, 6);
        f.factors.insert(11, 2);
        f.factors.insert(13, 3);
        f.factors.insert(17, 1);
        f.factors.insert(19, 1);
        f.factors.insert(23, 1);
        f.factors.insert(29, 1);
        f.factors.insert(31, 1);
        f.factors.insert(41, 1);
        f.factors.insert(47, 1);
        f.factors.insert(59, 1);
        f.factors.insert(71, 1);
        f
    }
    
    fn to_value(&self) -> BigUint {
        let mut result = BigUint::one();
        for (&prime, &exp) in &self.factors {
            result *= BigUint::from(prime).pow(exp);
        }
        result
    }
    
    fn format(&self) -> String {
        let mut parts: Vec<_> = self.factors.iter()
            .filter(|(_, &exp)| exp > 0)
            .collect();
        parts.sort_by_key(|(p, _)| **p);
        
        parts.iter()
            .map(|(&p, &exp)| {
                if exp == 1 {
                    format!("{}", p)
                } else {
                    format!("{}^{}", p, exp)
                }
            })
            .collect::<Vec<_>>()
            .join(" × ")
    }
}

/// Walk step
#[derive(Debug, Clone)]
struct WalkStep {
    step_num: u8,
    removed: PrimeFactorization,
    remaining: PrimeFactorization,
    value: BigUint,
    base: u32,
    representation: Vec<u8>,
}

/// Convert to base
fn to_base(n: &BigUint, base: u32) -> Vec<u8> {
    if n.is_zero() {
        return vec![0];
    }
    
    let mut digits = Vec::new();
    let mut num = n.clone();
    let base_big = BigUint::from(base);
    
    while !num.is_zero() {
        let digit = (&num % &base_big).to_u32().unwrap_or(0) as u8;
        digits.push(digit);
        num /= &base_big;
    }
    
    digits.reverse();
    digits
}

/// Generate walk for given base
fn walk_in_base(base: u32) -> Vec<WalkStep> {
    let monster = PrimeFactorization::monster();
    
    // Step 2: Remove 17, 59
    let mut step2_removed = PrimeFactorization::new();
    step2_removed.factors.insert(17, 1);
    step2_removed.factors.insert(59, 1);
    
    let mut step2_remaining = monster.clone();
    step2_remaining.factors.remove(&17);
    step2_remaining.factors.remove(&59);
    
    // Step 4: Remove 8 factors (Group 1)
    let mut step4_removed = PrimeFactorization::new();
    step4_removed.factors.insert(7, 6);
    step4_removed.factors.insert(11, 2);
    step4_removed.factors.insert(17, 1);
    step4_removed.factors.insert(19, 1);
    step4_removed.factors.insert(29, 1);
    step4_removed.factors.insert(31, 1);
    step4_removed.factors.insert(41, 1);
    step4_removed.factors.insert(59, 1);
    
    let mut step4_remaining = monster.clone();
    for (&p, _) in &step4_removed.factors {
        step4_remaining.factors.remove(&p);
    }
    
    // Step 6: Remove 4 factors (Group 2)
    let mut step6_removed = PrimeFactorization::new();
    step6_removed.factors.insert(3, 20);
    step6_removed.factors.insert(5, 9);
    step6_removed.factors.insert(13, 3);
    step6_removed.factors.insert(31, 1);
    
    let mut step6_remaining = monster.clone();
    for (&p, _) in &step6_removed.factors {
        step6_remaining.factors.remove(&p);
    }
    
    // Step 8: Remove 4 factors (Group 3)
    let mut step8_removed = PrimeFactorization::new();
    step8_removed.factors.insert(3, 20);
    step8_removed.factors.insert(13, 3);
    step8_removed.factors.insert(31, 1);
    step8_removed.factors.insert(71, 1);
    
    let mut step8_remaining = monster.clone();
    for (&p, _) in &step8_removed.factors {
        step8_remaining.factors.remove(&p);
    }
    
    // Step 10: Just 71
    let mut step10_remaining = PrimeFactorization::new();
    step10_remaining.factors.insert(71, 1);
    
    vec![
        WalkStep {
            step_num: 1,
            removed: PrimeFactorization::new(),
            remaining: monster.clone(),
            value: monster.to_value(),
            base,
            representation: to_base(&monster.to_value(), base),
        },
        WalkStep {
            step_num: 2,
            removed: step2_removed,
            remaining: step2_remaining.clone(),
            value: step2_remaining.to_value(),
            base,
            representation: to_base(&step2_remaining.to_value(), base),
        },
        WalkStep {
            step_num: 4,
            removed: step4_removed,
            remaining: step4_remaining.clone(),
            value: step4_remaining.to_value(),
            base,
            representation: to_base(&step4_remaining.to_value(), base),
        },
        WalkStep {
            step_num: 6,
            removed: step6_removed,
            remaining: step6_remaining.clone(),
            value: step6_remaining.to_value(),
            base,
            representation: to_base(&step6_remaining.to_value(), base),
        },
        WalkStep {
            step_num: 8,
            removed: step8_removed,
            remaining: step8_remaining.clone(),
            value: step8_remaining.to_value(),
            base,
            representation: to_base(&step8_remaining.to_value(), base),
        },
        WalkStep {
            step_num: 10,
            removed: monster,
            remaining: step10_remaining.clone(),
            value: BigUint::from(71u32),
            base,
            representation: to_base(&BigUint::from(71u32), base),
        },
    ]
}

fn main() {
    println!("🔢 Monster Walk with Primes - All Bases");
    println!("========================================\n");
    
    // Show walk in key bases
    for base in [2, 10, 16, 71] {
        println!("Base {}:", base);
        let walk = walk_in_base(base);
        
        for step in &walk {
            println!("  Step {}: {} digits",
                step.step_num,
                step.representation.len()
            );
            if step.step_num == 4 && base == 10 {
                println!("    Primes: {}", step.remaining.format());
                println!("    ✓ Preserves 8080");
            }
        }
        println!();
    }
    
    // Generate all bases
    println!("Generating all bases 2-71...");
    let all_walks: Vec<_> = (2..=71)
        .map(|base| (base, walk_in_base(base)))
        .collect();
    
    println!("✅ Generated {} bases", all_walks.len());
    println!("✅ Each walk has {} steps", all_walks[0].1.len());
}
